osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -380,22 +387,25 @@ private void update(BeatmapSetInfo set, BeatmapInfo beatmap)
 
                 var req = new GetBeatmapRequest(beatmap);
 
-                req.Success += res =>
+                req.Failure += e => { LogForModel(set, $"Online retrieval failed for {beatmap} ({e.Message})"); };
+
+                try
                 {
-                    LogForModel(set, $"Online retrieval mapped {beatmap} to {res.OnlineBeatmapSetID} / {res.OnlineBeatmapID}.");
+                    // intentionally blocking to limit web request concurrency
+                    req.Perform(api);
+
+                    var res = req.Result;
 
                     beatmap.Status = res.Status;
                     beatmap.BeatmapSet.Status = res.BeatmapSet.Status;
                     beatmap.BeatmapSet.OnlineBeatmapSetID = res.OnlineBeatmapSetID;
-                    beatmap.OnlineBeatmapID = res.OnlineBeatmapID;
-                };
 
-                req.Failure += e => { LogForModel(set, $"Online retrieval failed for {beatmap} ({e.Message})"); };
+                    // note that this check only needs to be here if two identical hashed beatmaps exist int he same import.
+                    // probably fine to leave it for safety.
+                    if (set.Beatmaps.All(b => b.OnlineBeatmapID != res.OnlineBeatmapID))
```

Comment:
Just documenting although we've discussed this IRL.

This is going to fail if a bad beatmap in the set already contains the online beatmap ID of another beatmap in the set.